### PR TITLE
Custom CA

### PR DESCRIPTION
**Pull request by chess-knight**
This PR addresses epic #372 - issues #373 #374 #375 #376 in that epic.

Edit: Issue #488 created - document CA rotation - will be addressed in separate PR

Terraform automatically detects *cacert* in *clouds.yaml* and transfers it to the mgmt host so openstack commands work.
Then cacert is included via OPENSTACK_CLOUD_CACERT_B64 to the CAPO controller so CAPO can create an infrastructure for the workload cluster.
Finally, cacert is also transferred to the master and worker nodes to the "/etc/ssl/certs" destination and then OCCM and CCSI can pick it up via hostPath mounts. There is no action for OCCM as the mounts are already present in the [manifest](https://github.com/kubernetes/cloud-provider-openstack/blob/master/manifests/controller-manager/openstack-cloud-controller-manager-ds.yaml), but for the CCSI we need to create these mounts (it is done by *yq*).

**How I tested custom CA**
I used [CAPO](https://github.com/kubernetes-sigs/cluster-api-provider-openstack) for spawning devstack on gx-scs OpenStack.
See [docs](https://cluster-api-openstack.sigs.k8s.io/development/development.html#openstack) for CAPO e2e testing.
Follow these steps to create devstack with a custom CA:
1. create ssh key
2. my environment looks like this:
   ```
   export RESOURCE_TYPE="openstack"
   export OS_CLOUD=gx-scs
   export OPENSTACK_FLAVOR_controller=SCS-16V:64
   export OPENSTACK_FLAVOR_worker=SCS-8V:32
   export OPENSTACK_PUBLIC_NETWORK=ext01
   export SSH_PUBLIC_KEY_FILE=</path/to/the/public_key>
   export SSH_PRIVATE_KEY_FILE=</path/to/the/private_key>

   export OPENSTACK_ADDITIONAL_SERVICES=,tls-proxy
   ```
3. Service `tls-proxy` will spawn devstack with tls support, and it also generates cacert
4. `./hack/ci/create_devstack.sh`
   it will create openstack controller and worker. When tls-proxy is used worker will
   not connect to controller because of ssl problems and I was not able to configure scripts
   in that way that worker will successfully connect. But the controller is enough for these purposes.
5. cleanup can be done via `./hack/ci/create_devstack.sh cleanup`

Follow these steps to test k8s-cluster-api-provider with devstack custom CA:
1. controller has public IP, so it is possible to connect via e.g. `ssh -i ${SSH_PRIVATE_KEY_FILE} cloud@213.131.230.213`
2. `git clone https://github.com/SovereignCloudStack/k8s-cluster-api-provider.git -b custom_ca`
3. `cd k8s-cluster-api-provider/terraform/`
4. `sudo snap install yq`
5. `sudo snap install terraform --classic`
6. create `environments/environment-devstack.tfvars`:
   ```terraform
   cloud_provider    = "devstack"
   availability_zone = "testaz1"
   external          = "public"
   kind_flavor       = "m1.small"
   controller_flavor = "m1.small"
   worker_flavor     = "m1.small"
   dns_nameservers   = ["62.138.222.111", "62.138.222.222"]
   worker_count = 2
   prefix = "devstack"
   testcluster_name = "devstackcluster"
   image = "focal-server-cloudimg-amd64"
   kube_image_raw = false
   ```
7. Add `interface: public` to the `/etc/openstack/clouds.yaml` *cloud* **devstack**, `cacert` can be also seen there
8. comment *availability_zone_hints* in [neutron.tf](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/terraform/neutron.tf#L76)
9. `export OS_CLOUD=devstack`
10. `make create`
11. Creation of management and workload cluster should be successful, and later it can be tested e.g. with `make check-conformance`
12. `make fullclean`

### CODE REVIEW

**Comment by jschoone** [comment]
@chess-knight I was finally able to test all these steps. Worked like charm, thank you for the clear guide, this is really cool especially that you provide me a way to test the repo on DevStack :smile: 

I checked the CAs I had in the environment and based on them I would approve this PR, but stillt want to go through the code to have an even better understanding of it.

> **How I tested custom CA** I used [CAPO](https://github.com/kubernetes-sigs/cluster-api-provider-openstack) for spawning devstack on gx-scs OpenStack. See [docs](https://cluster-api-openstack.sigs.k8s.io/development/development.html#openstack) for CAPO e2e testing. Follow these steps to create devstack with a custom CA:
> 
> 1. create ssh key
> 2. my environment looks like this:
>    ```
>    export RESOURCE_TYPE="openstack"
>    export OS_CLOUD=gx-scs
>    export OPENSTACK_FLAVOR_controller=SCS-16V:64
>    export OPENSTACK_FLAVOR_worker=SCS-8V:32
>    export OPENSTACK_PUBLIC_NETWORK=ext01
>    export SSH_PUBLIC_KEY_FILE=</path/to/the/public_key>
>    export SSH_PRIVATE_KEY_FILE=</path/to/the/private_key>
>    
>    export OPENSTACK_ADDITIONAL_SERVICES=,tls-proxy
>    ```
> 3. Service `tls-proxy` will spawn devstack with tls support, and it also generates cacert
> 4. `./hack/ci/create_devstack.sh`
>    it will create openstack controller and worker. When tls-proxy is used worker will
>    not connect to controller because of ssl problems and I was not able to configure scripts
>    in that way that worker will successfully connect. But the controller is enough for these purposes.
> 5. cleanup can be done via `./hack/ci/create_devstack.sh cleanup`
> 
> Follow these steps to test k8s-cluster-api-provider with devstack custom CA:
> 
> 1. controller has public IP, so it is possible to connect via e.g. `ssh -i ${SSH_PRIVATE_KEY_FILE} cloud@213.131.230.213`
> 2. `git clone https://github.com/SovereignCloudStack/k8s-cluster-api-provider.git -b custom_ca`
> 3. `cd k8s-cluster-api-provider/terraform/`
> 4. `sudo snap install yq`
> 5. `sudo snap install terraform --classic`
> 6. create `environments/environment-devstack.tfvars`:
>    ```terraform
>    cloud_provider    = "devstack"
>    availability_zone = "testaz1"
>    external          = "public"
>    kind_flavor       = "m1.small"
>    controller_flavor = "m1.small"
>    worker_flavor     = "m1.small"
>    dns_nameservers   = ["62.138.222.111", "62.138.222.222"]
>    worker_count = 2
>    prefix = "devstack"
>    testcluster_name = "devstackcluster"
>    image = "focal-server-cloudimg-amd64"
>    kube_image_raw = false
>    ```
> 7. Add `interface: public` to the `/etc/openstack/clouds.yaml` _cloud_ **devstack**, `cacert` can be also seen there
> 8. comment _availability_zone_hints_ in [neutron.tf](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/terraform/neutron.tf#L76)
> 9. `export OS_CLOUD=devstack`
> 10. `make create`
> 11. Creation of management and workload cluster should be successful, and later it can be tested e.g. with `make check-conformance`
> 12. `make fullclean`

**Review by matofeder — APPROVED** [review]
I added just a few minor suggestions in the docs section.
LGTM as this PR has been already successfully tested by @jschoone

File: doc/usage/custom-ca.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Custom CA
+
+Communication with OpenStack API should be protected by a certificate.
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply doc/usage/custom-ca.md]: Addressed in b3e81c1037e4290743bbc1eefbc043dca875a913

File: doc/usage/custom-ca.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+This certificate is usually issued by public CA but sometimes the custom or private CA is used.
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply doc/usage/custom-ca.md]: Addressed in b3e81c1037e4290743bbc1eefbc043dca875a913

File: doc/usage/custom-ca.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    auth:
+      auth_url: https://10.0.3.15/identity
+      project_domain_id: default
+      project_name: demo
+      user_domain_id: default
+    identity_api_version: 3
+    region_name: RegionOne
+    interface: public
+```
+Here the file `ca-bundle.pem` contains custom root CA and potentially intermediate CA.
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply doc/usage/custom-ca.md]: Addressed in b3e81c1037e4290743bbc1eefbc043dca875a913

File: doc/usage/custom-ca.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+> The `ca-bundle.pem` file will be copied to the management server and used by CAPO
+> in the management cluster. Also, it will be copied to the workload cluster(master and worker nodes)
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply doc/usage/custom-ca.md]: Addressed in b3e81c1037e4290743bbc1eefbc043dca875a913

File: doc/usage/custom-ca.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+> and mounted and used by OCCM and CCSI pods.
+> So provide only the necessary certificates in that file.
+
+Steps of what happens with the custom cacert in k8s-cluster-api-provider:
+1. `cacert` setting is provided inside clouds.yaml
+2. Cacert file is copied to the management server by terraform
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply doc/usage/custom-ca.md]: Addressed in b3e81c1037e4290743bbc1eefbc043dca875a913

File: doc/usage/custom-ca.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+3. During the management server bootstrap process
+   cacert is injected to the *~/cluster-defaults/cluster-template.yaml* to
+   *KubeadmControlPlane* and *KubeadmConfigTemplate* files, so it will be later
+   copied to the workload cluster(master and worker nodes) provisioned by CAPO
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply doc/usage/custom-ca.md]: Addressed in b3e81c1037e4290743bbc1eefbc043dca875a913

File: doc/usage/custom-ca.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+4. When the creation of the workload cluster starts, *~/cluster-defaults/cluster-template.yaml*
+   is copied into workload cluster directory(*~/$CLUSTER_NAME/*)
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply doc/usage/custom-ca.md]: Addressed in b3e81c1037e4290743bbc1eefbc043dca875a913

File: doc/usage/custom-ca.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+5. Then the cacert is base64 encoded and saved in OPENSTACK_CLOUD_CACERT_B64 variable
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply doc/usage/custom-ca.md]: Addressed in b3e81c1037e4290743bbc1eefbc043dca875a913

File: doc/usage/custom-ca.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+   inside *~/$CLUSTER_NAME/clusterctl.yaml*, so it can be used during
+   the workload cluster templating
+6. Later, when the workload cluster templates are applied to the management cluster,
+   secret with base64 encoded cacert is created and used by CAPO
+7. CAPO will create workload cluster(thanks to steps 5. and 6.) and cacert is
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply doc/usage/custom-ca.md]: Addressed in b3e81c1037e4290743bbc1eefbc043dca875a913

File: doc/usage/custom-ca.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+   transferred to the master and worker nodes(thanks to steps 3. and 4.).
```
[the remark this answers is not in the document]
Comment (chess-knight) [reply doc/usage/custom-ca.md]: Addressed in b3e81c1037e4290743bbc1eefbc043dca875a913

**Review by fdobrovolny — APPROVED** [review]

**Pull request merged**